web/src/main/java/org/springframework/security/web/savedrequest/CookieRequestCache.java

```diff
@@ -0,0 +1,130 @@
+import org.springframework.web.util.WebUtils;
+
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+
+/**
+ * {@code RequestCache} which stores the {@code SavedRequest} in a cookie.
+ *
+ * @author Zeeshan Adnan
+ */
+public class CookieRequestCache implements RequestCache {
+
+	private RequestMatcher requestMatcher = AnyRequestMatcher.INSTANCE;
+	protected final Log logger = LogFactory.getLog(this.getClass());
+
+	private String cookieName = "saved_request";
+	private String cookiePath = null;
+	private int cookieMaxAge = -1;
+
+	@Override
+	public void saveRequest(HttpServletRequest request, HttpServletResponse response) {
+		if (requestMatcher.matches(request)) {
+			String redirectUrl = UrlUtils.buildFullRequestUrl(request);
+			Cookie savedCookie = new Cookie(cookieName, redirectUrl);
+			savedCookie.setMaxAge(cookieMaxAge);
+			savedCookie.setSecure(request.isSecure());
+			savedCookie.setPath(cookiePath);
+			savedCookie.setHttpOnly(true);
+			response.addCookie(savedCookie);
+		} else {
+			logger.debug("Request not saved as configured RequestMatcher did not match");
+		}
+	}
+
+	@Override
+	public SavedRequest getRequest(HttpServletRequest request, HttpServletResponse response) {
+		Cookie savedRequestCookie = WebUtils.getCookie(request, cookieName);
+		return savedRequestCookie != null ? new SimpleSavedRequest(savedRequestCookie.getValue()) : null;
+	}
+
+	@Override
+	public HttpServletRequest getMatchingRequest(HttpServletRequest request, HttpServletResponse response) {
+		SavedRequest savedRequest = getRequest(request, response);
+		if (savedRequest != null) {
+			removeRequest(request, response);
+		}
+		return null;
+	}
+
+	@Override
+	public void removeRequest(HttpServletRequest request, HttpServletResponse response) {
+		Cookie removeSavedRequestCookie = new Cookie(cookieName, "");
+		removeSavedRequestCookie.setSecure(request.isSecure());
+		removeSavedRequestCookie.setHttpOnly(true);
+		removeSavedRequestCookie.setPath(cookiePath);
+		removeSavedRequestCookie.setMaxAge(0);
+		response.addCookie(removeSavedRequestCookie);
+	}
+
+	/**
+	 * Allows selective use of saved requests for a subset of requests. By default any
+	 * request will be cached by the {@code saveRequest} method.
+	 * <p>
+	 * If set, only matching requests will be cached.
+	 *
+	 * @param requestMatcher a request matching strategy which defines which requests
+	 *                       should be cached.
+	 */
+	public void setRequestMatcher(RequestMatcher requestMatcher) {
+		Assert.notNull(requestMatcher, "requestMatcher should not be null");
+		this.requestMatcher = requestMatcher;
+	}
+
+	/**
+	 * If the {@code cookieName} property is set, the request is stored in
+	 * the cookie using this attribute name. Default is
+	 * "saved_request".
+	 *
+	 * @param cookieName is the new name of the cookie.
+	 */
+	public void setCookieName(String cookieName) {
```

Comment:
I don't think we need to make the cookie name configurable.
We can always add this functionality later if a user requests it.
Similarly, I don't think we need `setCookieMaxAge` and `setCookiePath` for now.